s3auth-relay/src/main/java/com/s3auth/relay/AbstractFacade.java

```diff
@@ -0,0 +1,292 @@
+ * NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.s3auth.relay;
+
+import com.jcabi.aspects.Loggable;
+import com.jcabi.aspects.Tv;
+import com.jcabi.log.Logger;
+import com.jcabi.log.VerboseRunnable;
+import com.jcabi.log.VerboseThreads;
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.TimeUnit;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+
+/**
+ * Abstract facade (port listener).
+ *
+ * @author Yegor Bugayenko (yegor@tpc2.com)
+ * @author Simon Njenga (simtuje@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ */
+@ToString
+@EqualsAndHashCode(of = { "sockets", "server", "secured" })
+@SuppressWarnings({ "PMD.DoNotUseThreads", "PMD.TooManyMethods" })
+@Loggable(Loggable.DEBUG)
+abstract class AbstractFacade implements Closeable {
+
+    /**
+     * How many threads to use.
+     */
+    protected static final int THREADS = Tv.HUNDRED;
+
+    /**
+     * Executor service, with socket openers.
+     */
+    private transient ScheduledExecutorService frontend;
+
+    /**
+     * Executor service, with consuming threads.
+     */
+    private transient ScheduledExecutorService backend;
+
+    /**
+     * Blocking queue of ready-to-be-processed sockets.
+     */
+    private final transient BlockingQueue<Socket> sockets =
+        new SynchronousQueue<Socket>();
+
+    /**
+     * Server socket.
+     */
+    private transient ServerSocket server;
+
+    /**
+     * Secured Server socket.
+     */
+    private transient ServerSocket secured;
+
+    @Override
+    public void close() throws IOException {
+        try {
+            this.shutdown(this.getFrontend());
+            this.shutdown(this.getBackend());
+        } catch (final InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            throw new IOException(ex);
+        }
+        this.getServer().close();
+        this.getSecured().close();
+    }
+
+    /**
+     * Get frontend.
+     * @return The frontend
+     */
+    protected final ScheduledExecutorService getFrontend() {
```

Comment:
@simonjenga no getters/setters. use a constructor and set the members as final to keep it immutable.